File: contracts/wasm/inccounter/rs/inccounterimpl/src/funcs.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use wasmlib::*;
use inccounter::*;
use crate::*;

const HEX: &str = "0123456789abcdef";

static mut LOCAL_STATE_MUST_INCREMENT: bool = false;

pub fn func_init(_ctx: &ScFuncContext, f: &InitContext) {
    if f.params.counter().exists() {
        let counter = f.params.counter().value();
        f.state.counter().set_value(counter);
    }
}

pub fn func_call_increment(ctx: &ScFuncContext, f: &CallIncrementContext) {
    let counter = f.state.counter();
    let value = counter.value();
    counter.set_value(value + 1);
    if value == 0 {
        ScFuncs::call_increment(ctx).func.call();
    }
}

pub fn func_call_increment_recurse5x(ctx: &ScFuncContext, f: &CallIncrementRecurse5xContext) {
    let counter = f.state.counter();
    let value = counter.value();
    counter.set_value(value + 1);
    if value < 5 {
        ScFuncs::call_increment_recurse5x(ctx).func.call();
    }
}

pub fn func_endless_loop(_ctx: &ScFuncContext, _f: &EndlessLoopContext) {
    loop {}
}

pub fn func_increment(_ctx: &ScFuncContext, f: &IncrementContext) {
    let counter = f.state.counter();
    counter.set_value(counter.value() + 1);
}

pub fn func_increment_with_delay(ctx: &ScFuncContext, f: &IncrementWithDelayContext) {
    let delay = f.params.delay().value();
    let inc = ScFuncs::call_increment(ctx);
    inc.func.delay(delay).post();
}

pub fn func_local_state_internal_call(ctx: &ScFuncContext, f: &LocalStateInternalCallContext) {
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = false;
    }
    when_must_increment_state(ctx, &f.state);
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = true;
    }
    when_must_increment_state(ctx, &f.state);
    when_must_increment_state(ctx, &f.state);
    // counter ends up as 2
}

pub fn func_local_state_post(ctx: &ScFuncContext, _f: &LocalStatePostContext) {
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = false;
    }
    // prevent multiple identical posts, need a dummy param to differentiate them
    local_state_post(ctx, 1);
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = true;
    }
    local_state_post(ctx, 2);
    local_state_post(ctx, 3);
    // counter ends up as 0
}

pub fn func_local_state_sandbox_call(ctx: &ScFuncContext, _f: &LocalStateSandboxCallContext) {
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = false;
    }
    ScFuncs::when_must_increment(ctx).func.call();
    unsafe {
        LOCAL_STATE_MUST_INCREMENT = true;
    }
    ScFuncs::when_must_increment(ctx).func.call();
    ScFuncs::when_must_increment(ctx).func.call();
    // counter ends up as 0
}

pub fn func_post_increment(ctx: &ScFuncContext, f: &PostIncrementContext) {
    let counter = f.state.counter();
    let value = counter.value();
    counter.set_value(value + 1);
    if value == 0 {
        ScFuncs::increment(ctx).func.post();
    }
}

pub fn func_repeat_many(ctx: &ScFuncContext, f: &RepeatManyContext) {
    let counter = f.state.counter();
    let value = counter.value();
    counter.set_value(value + 1);
    let state_repeats = f.state.num_repeats();
    let mut repeats = f.params.num_repeats().value();
    if repeats == 0 {
        repeats = state_repeats.value();
        if repeats == 0 {
            return;
        }
    }
    state_repeats.set_value(repeats - 1);
    ScFuncs::repeat_many(ctx).func.post();
}

pub fn func_test_vli_codec(ctx: &ScFuncContext, _f: &TestVliCodecContext) {
    vli_save(ctx, "v-129", -129);
    vli_save(ctx, "v-128", -128);
    vli_save(ctx, "v-127", -127);
    vli_save(ctx, "v-126", -126);
    vli_save(ctx, "v-65", -65);
    vli_save(ctx, "v-64", -64);
    vli_save(ctx, "v-63", -63);
    vli_save(ctx, "v-62", -62);
    vli_save(ctx, "v-2", -2);
    vli_save(ctx, "v-1", -1);
    vli_save(ctx, "v 0", 0);
    vli_save(ctx, "v+1", 1);
    vli_save(ctx, "v+2", 2);
    vli_save(ctx, "v+62", 62);
    vli_save(ctx, "v+63", 63);
    vli_save(ctx, "v+64", 64);
    vli_save(ctx, "v+65", 65);
    vli_save(ctx, "v+126", 126);
    vli_save(ctx, "v+127", 127);
    vli_save(ctx, "v+128", 128);
    vli_save(ctx, "v+129", 129);
}

pub fn func_test_vlu_codec(ctx: &ScFuncContext, _f: &TestVluCodecContext) {
    vlu_save(ctx, "v 0", 0);
    vlu_save(ctx, "v+1", 1);
    vlu_save(ctx, "v+2", 2);
    vlu_save(ctx, "v+62", 62);
    vlu_save(ctx, "v+63", 63);
    vlu_save(ctx, "v+64", 64);
    vlu_save(ctx, "v+65", 65);
    vlu_save(ctx, "v+126", 126);
    vlu_save(ctx, "v+127", 127);
    vlu_save(ctx, "v+128", 128);
    vlu_save(ctx, "v+129", 129);
}

pub fn func_when_must_increment(ctx: &ScFuncContext, f: &WhenMustIncrementContext) {
    when_must_increment_state(ctx, &f.state);
}

// note that get_counter mirrors the state of the 'counter' state variable
// which means that if the state variable was not present it also will not be present in the result
pub fn view_get_counter(_ctx: &ScViewContext, f: &GetCounterContext) {
    let counter = f.state.counter();
    if counter.exists() {
        f.results.counter().set_value(counter.value());
    }
}

pub fn view_get_vli(_ctx: &ScViewContext, f: &GetVliContext) {
    let mut enc = WasmEncoder::new();
    let n = f.params.ni64().value();
    int64_encode(&mut enc,n);
    let buf = enc.buf();
    let mut dec = WasmDecoder::new(&buf);
    let x = int64_decode(&mut dec);

    let mut str = n.to_string() + " -";
    for b in &buf {
        let h1 = ((b >> 4) & 0x0f) as usize;
        let h2 = (b & 0x0f) as usize;
        str += &(" ".to_string() + &HEX[h1..h1+1] + &HEX[h2..h2+1]);
    }
    str += &(" - ".to_string() + &x.to_string());

    f.results.ni64().set_value(n);
    f.results.xi64().set_value(x);
    f.results.str().set_value(&str);
    f.results.buf().set_value(&buf);
}

pub fn view_get_vlu(_ctx: &ScViewContext, f: &GetVluContext) {
    let mut enc = WasmEncoder::new();
    let n = f.params.nu64().value();
    uint64_encode(&mut enc, n);
    let buf = enc.buf();
    let mut dec = WasmDecoder::new(&buf);
    let x = uint64_decode(&mut dec);

    let mut str = n.to_string() + " -";
    for b in &buf {
        let h1 = ((b >> 4) & 0x0f) as usize;
        let h2 = (b & 0x0f) as usize;
        str += &(" ".to_string() + &HEX[h1..h1+1] + &HEX[h2..h2+1]);
    }
    str += &(" - ".to_string() + &x.to_string());

    f.results.nu64().set_value(n);
    f.results.xu64().set_value(x);
    f.results.str().set_value(&str);
    f.results.buf().set_value(&buf);
}

//////////////////////////////// util funcs \\\\\\\\\\\\\\\\\\\\\\\\\\\\\

fn local_state_post(ctx: &ScFuncContext, nr: i64) {
    //note: we add a dummy parameter here to prevent "duplicate outputs not allowed" error
    let f = ScFuncs::when_must_increment(ctx);
    f.params.dummy().set_value(nr);
    f.func.post();
}

fn vli_save(ctx: &ScFuncContext, name: &str, value: i64) {
    let mut enc = WasmEncoder::new();
    let state = ctx.raw_state();
    let key = string_to_bytes(name);
    state.set(&key, &enc.vli_encode(value).buf());

    let buf = state.get(&key);
    let mut dec = WasmDecoder::new(&buf);
    let val = dec.vli_decode(64);
    if val != value {
        ctx.log(&(name.to_string() + " in : " + &value.to_string()));
        ctx.log(&(name.to_string() + " out: " + &val.to_string()));
    }
}

fn vlu_save(ctx: &ScFuncContext, name: &str, value: u64) {
    let mut enc = WasmEncoder::new();
    let state = ctx.raw_state();
    let key = string_to_bytes(name);
    state.set(&key, &enc.vlu_encode(value).buf());

    let buf = state.get(&key);
    let mut dec = WasmDecoder::new(&buf);
    let val = dec.vlu_decode(64);
    if val != value {
        ctx.log(&(name.to_string() + " in : " + &value.to_string()));
        ctx.log(&(name.to_string() + " out: " + &val.to_string()));
    }
}

fn when_must_increment_state(ctx: &ScFuncContext, state: &MutableIncCounterState) {
    ctx.log("when_must_increment called");
    unsafe {
        if !LOCAL_STATE_MUST_INCREMENT {
            return;
        }
    }
    let counter = state.counter();
    counter.set_value(counter.value() + 1);
}

File: contracts/wasm/inccounter/ts/inccounterimpl/funcs.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from "wasmlib"
import * as wasmtypes from "wasmlib/wasmtypes";
import * as sc from "../inccounter/index";

const hex = "0123456789abcdef";

let localStateMustIncrement: boolean = false;

export function funcInit(ctx: wasmlib.ScFuncContext, f: sc.InitContext): void {
    if (f.params.counter().exists()) {
        let counter = f.params.counter().value();
        f.state.counter().setValue(counter);
    }
}

export function funcCallIncrement(ctx: wasmlib.ScFuncContext, f: sc.CallIncrementContext): void {
    let counter = f.state.counter();
    let value = counter.value();
    counter.setValue(value + 1);
    if (value == 0) {
        sc.ScFuncs.callIncrement(ctx).func.call();
    }
}

export function funcCallIncrementRecurse5x(ctx: wasmlib.ScFuncContext, f: sc.CallIncrementRecurse5xContext): void {
    let counter = f.state.counter();
    let value = counter.value();
    counter.setValue(value + 1);
    if (value < 5) {
        sc.ScFuncs.callIncrementRecurse5x(ctx).func.call();
    }
}

export function funcEndlessLoop(ctx: wasmlib.ScFuncContext, f: sc.EndlessLoopContext): void {
    for (; ;) {
    }
}

export function funcIncrement(ctx: wasmlib.ScFuncContext, f: sc.IncrementContext): void {
    let incValue: i64 = 1;
    let param = f.params.counter();
    if (param.exists()) {
        incValue = param.value()
    }

    let counter = f.state.counter();
    counter.setValue(counter.value() + incValue);
}

export function funcIncrementWithDelay(ctx: wasmlib.ScFuncContext, f: sc.IncrementWithDelayContext): void {
    let delay = f.params.delay().value();
    let inc = sc.ScFuncs.callIncrement(ctx);
    inc.func.delay(delay).post();
}

export function funcLocalStateInternalCall(ctx: wasmlib.ScFuncContext, f: sc.LocalStateInternalCallContext): void {
    localStateMustIncrement = false;
    whenMustIncrementState(ctx, f.state);
    localStateMustIncrement = true;
    whenMustIncrementState(ctx, f.state);
    whenMustIncrementState(ctx, f.state);
    // counter ends up as 2
}

export function funcLocalStatePost(ctx: wasmlib.ScFuncContext, f: sc.LocalStatePostContext): void {
    localStateMustIncrement = false;
    // prevent multiple identical posts, need a dummy param to differentiate them
    localStatePost(ctx, 1);
    localStateMustIncrement = true;
    localStatePost(ctx, 2);
    localStatePost(ctx, 3);
    // counter ends up as 0
}

export function funcLocalStateSandboxCall(ctx: wasmlib.ScFuncContext, f: sc.LocalStateSandboxCallContext): void {
    localStateMustIncrement = false;
    sc.ScFuncs.whenMustIncrement(ctx).func.call();
    localStateMustIncrement = true;
    sc.ScFuncs.whenMustIncrement(ctx).func.call();
    sc.ScFuncs.whenMustIncrement(ctx).func.call();
    // counter ends up as 0
}

export function funcPostIncrement(ctx: wasmlib.ScFuncContext, f: sc.PostIncrementContext): void {
    let counter = f.state.counter();
    let value = counter.value();
    counter.setValue(value + 1);
    if (value == 0) {
        sc.ScFuncs.increment(ctx).func.post();
    }
}

export function funcRepeatMany(ctx: wasmlib.ScFuncContext, f: sc.RepeatManyContext): void {
    let counter = f.state.counter();
    let value = counter.value();
    counter.setValue(value + 1);
    let stateRepeats = f.state.numRepeats();
    let repeats = f.params.numRepeats().value();
    if (repeats == 0) {
        repeats = stateRepeats.value();
        if (repeats == 0) {
            return;
        }
    }
    stateRepeats.setValue(repeats - 1);
    sc.ScFuncs.repeatMany(ctx).func.post();
}

export function funcTestVliCodec(ctx: wasmlib.ScFuncContext, f: sc.TestVliCodecContext): void {
    for (let i: i64 = -1600; i < 1600; i++) {
        let enc = new wasmtypes.WasmEncoder();
        wasmtypes.int64Encode(enc, i);
        let buf = enc.buf();
        // let txt = i.toString() + " -";
        // for (let j = 0; j < buf.length; j++) {
        //     let b = buf[j];
        //     txt += " " + hex.charAt((b >> 4) & 0x0f) + hex.charAt(b & 0x0f);
        // }
        let dec = new wasmtypes.WasmDecoder(buf);
        let v = wasmtypes.int64Decode(dec);
        // txt += " - " + v.toString();
        // ctx.log(txt);
        ctx.require(i == v, "coder value mismatch")
    }

    vliSave(ctx, "v-129", -129);
    vliSave(ctx, "v-128", -128);
    vliSave(ctx, "v-127", -127);
    vliSave(ctx, "v-126", -126);
    vliSave(ctx, "v-65", -65);
    vliSave(ctx, "v-64", -64);
    vliSave(ctx, "v-63", -63);
    vliSave(ctx, "v-62", -62);
    vliSave(ctx, "v-2", -2);
    vliSave(ctx, "v-1", -1);
    vliSave(ctx, "v 0", 0);
    vliSave(ctx, "v+1", 1);
    vliSave(ctx, "v+2", 2);
    vliSave(ctx, "v+62", 62);
    vliSave(ctx, "v+63", 63);
    vliSave(ctx, "v+64", 64);
    vliSave(ctx, "v+65", 65);
    vliSave(ctx, "v+126", 126);
    vliSave(ctx, "v+127", 127);
    vliSave(ctx, "v+128", 128);
    vliSave(ctx, "v+129", 129);
}

export function funcTestVluCodec(ctx: wasmlib.ScFuncContext, f: sc.TestVluCodecContext): void {
    for (let i: u64 = 0; i < 3200; i++) {
        let enc = new wasmtypes.WasmEncoder();
        wasmtypes.uint64Encode(enc, i);
        let buf = enc.buf();
        // let txt = i.toString() + " -";
        // for (let j = 0; j < buf.length; j++) {
        //     let b = buf[j];
        //     txt += " " + hex.charAt((b >> 4) & 0x0f) + hex.charAt(b & 0x0f);
        // }
        let dec = new wasmtypes.WasmDecoder(buf);
        let v = wasmtypes.uint64Decode(dec);
        // txt += " - " + v.toString();
        // ctx.log(txt);
        ctx.require(i == v, "coder value mismatch")
    }

    vluSave(ctx, "v 0", 0);
    vluSave(ctx, "v+1", 1);
    vluSave(ctx, "v+2", 2);
    vluSave(ctx, "v+62", 62);
    vluSave(ctx, "v+63", 63);
    vluSave(ctx, "v+64", 64);
    vluSave(ctx, "v+65", 65);
    vluSave(ctx, "v+126", 126);
    vluSave(ctx, "v+127", 127);
    vluSave(ctx, "v+128", 128);
    vluSave(ctx, "v+129", 129);
}

export function funcWhenMustIncrement(ctx: wasmlib.ScFuncContext, f: sc.WhenMustIncrementContext): void {
    whenMustIncrementState(ctx, f.state);
}

// note that getCounter mirrors the state of the 'counter' state variable
// which means that if the state variable was not present it also will not be present in the result
export function viewGetCounter(ctx: wasmlib.ScViewContext, f: sc.GetCounterContext): void {
    let counter = f.state.counter();
    if (counter.exists()) {
        f.results.counter().setValue(counter.value());
    }
}

export function viewGetVli(ctx: wasmlib.ScViewContext, f: sc.GetVliContext): void {
    let enc = new wasmtypes.WasmEncoder();
    let n = f.params.ni64().value();
    wasmtypes.int64Encode(enc, n);
    let buf = enc.buf();
    let dec = new wasmtypes.WasmDecoder(buf);
    let x = wasmtypes.int64Decode(dec);

    let str = n.toString() + " -";
    for (let j = 0; j < buf.length; j++) {
        let b = buf[j];
        str += " " + hex.charAt((b >> 4) & 0x0f) + hex.charAt(b & 0x0f);
    }
    str += " - " + x.toString();

    f.results.ni64().setValue(n);
    f.results.xi64().setValue(x);
    f.results.str().setValue(str);
    f.results.buf().setValue(buf);
}

export function viewGetVlu(ctx: wasmlib.ScViewContext, f: sc.GetVluContext): void {
    let enc = new wasmtypes.WasmEncoder();
    let n = f.params.nu64().value();
    wasmtypes.uint64Encode(enc, n);
    let buf = enc.buf();
    let dec = new wasmtypes.WasmDecoder(buf);
    let x = wasmtypes.uint64Decode(dec);

    let str = n.toString() + " -";
    for (let j = 0; j < buf.length; j++) {
        let b = buf[j];
        str += " " + hex.charAt((b >> 4) & 0x0f) + hex.charAt(b & 0x0f);
    }
    str += " - " + x.toString();

    f.results.nu64().setValue(n);
    f.results.xu64().setValue(x);
    f.results.str().setValue(str);
    f.results.buf().setValue(buf);
}

//////////////////////////////// util funcs \\\\\\\\\\\\\\\\\\\\\\\\\\\\\

function localStatePost(ctx: wasmlib.ScFuncContext, nr: i64): void {
    //note: we add a dummy parameter here to prevent "duplicate outputs not allowed" error
    let f = sc.ScFuncs.whenMustIncrement(ctx);
    f.params.dummy().setValue(nr);
    f.func.post();
}

function vliSave(ctx: wasmlib.ScFuncContext, name: string, value: i64): void {
    const enc = new wasmtypes.WasmEncoder();
    const state = ctx.rawState();
    const key = wasmtypes.stringToBytes(name);
    state.set(key, enc.vliEncode(value).buf());

    const buf = state.get(key);
    const dec = new wasmtypes.WasmDecoder(buf);
    const val = dec.vliDecode(64);
    if (val != value) {
        ctx.log(name.toString() + " in : " + value.toString());
        ctx.log(name.toString() + " out: " + val.toString());
    }
}

function vluSave(ctx: wasmlib.ScFuncContext, name: string, value: u64): void {
    const enc = new wasmtypes.WasmEncoder();
    const state = ctx.rawState();
    const key = wasmtypes.stringToBytes(name);
    state.set(key, enc.vluEncode(value).buf());

    const buf = state.get(key);
    const dec = new wasmtypes.WasmDecoder(buf);
    const val = dec.vluDecode(64);
    if (val != value) {
        ctx.log(name.toString() + " in : " + value.toString());
        ctx.log(name.toString() + " out: " + val.toString());
    }
}

function whenMustIncrementState(ctx: wasmlib.ScFuncContext, state: sc.MutableIncCounterState): void {
    ctx.log("whenMustIncrement called");
    if (!localStateMustIncrement) {
        return;
    }
    let counter = state.counter();
    counter.setValue(counter.value() + 1);
    ctx.log("whenMustIncrement incremented");
}
